### DISCUSSION TITLE
Unlock Hook @exporter side

### DISCUSSION THREAD

**Discussion by metalinjection** [labgrid-project/labgrid#1492] [Q&A] [answered]
Hi there,
I am using langurs now for some time however I am not thru all the code sections yet.
I was wondering if there is a kind of an „Unlock“ hook.
This way we could reset the device in a certain state after the person unlocks the place.
E.g. with the new lcus relays we could make sure that they are all reset to off state.

when I looked for hooks I only found the pytest hooks however there is code for the _unlock however I am not sure if this related to the client, coord or exporter?

thanks for any answer.

**Comment by Emantor** [comment]
As soon as you unlock the place, the user can't do any action on the place any longer, so by definition this won't work with the current design. I'd also like to avoid introducing the complexity to labgrid in the unlock step, I'd much rather have this solved via external scripting.

For CI loops this is straightforward, before unlocking the device, reset it to a known good state. For interactive use this is more tricky, I don't have an immediate idea yet.

**Reply by metalinjection** [reply]
Hi, I think I might expressed myself wrong here - this is not on the User end.
I am looking for a hook on the exporter side - this way, I could implement the "rollback" of this particular system.
For example if the system is in a certain state, I would like to make sure it is being e.g. reflashed with a valid firmware, certain files are being checked and possibly reset to the default state, aso. - literally tidy up the "workplace".

So, how does the exporter know that the user "has left the building"? I mean I guess the exporter does the choosing of the ports for each object at that place or is that all done - every time - by the coordinator?

Sorry, if I ask this.. I guess I could go deeper into the jungle of labgrid land but I didnt manage to follow all trails in their yet ;-)

Cheers

**Reply by Emantor** [reply, marked as answer]
The exporter has no knowledge of a place as is, it only knows about resources a specific place acquired, so there is no way for the exporter to now when to run a hook. The place to resource mapping is only known on the coordinator side.

Labgrid has no idea when a user has left the building, we know when someone/something (a user, CI,…) has acquired a place, but we have no idea when who/whatever has acquired the place is done using it except for the release event, but at this point interaction with the target is no longer possible (since the resources are released).

**Reply by metalinjection** [reply]
Hi, ok… I get it… already talked with a friend that I might need to add a new Strategy than for Tidy Up
